quic/s2n-quic-platform/src/socket/io/tx.rs

```diff
@@ -0,0 +1,439 @@
+    is_full: bool,
+}
+
+impl<T: Message> Tx<T> {
+    #[inline]
+    pub fn new(channels: Vec<Producer<T>>, gso: Gso, max_mtu: MaxMtu) -> Self {
+        Self {
+            channels,
+            gso,
+            max_mtu: max_mtu.into(),
+            is_full: true,
+        }
+    }
+}
+
+impl<T: Message> tx::Tx for Tx<T> {
+    type PathHandle = T::Handle;
+    type Queue = TxQueue<'static, T>;
+    type Error = ();
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context) -> Poll<Result<(), Self::Error>> {
+        // We only need to poll for capacity if we completely filled up all of the channels.
+        // If we always polled, this would cause the endpoint to spin since most of the time it has
+        // capacity for sending.
+        if !self.is_full {
+            return Poll::Pending;
+        }
+
+        let mut is_any_ready = false;
+        let mut is_all_closed = true;
+
+        for channel in &mut self.channels {
+            match channel.poll_acquire(1, cx) {
+                Poll::Ready(_) => {
+                    is_all_closed = false;
+                    is_any_ready = true;
+                }
+                Poll::Pending => {
+                    is_all_closed &= !channel.is_open();
+                }
+            }
+        }
+
+        // if all of the channels were closed then shut the task down
+        if is_all_closed {
+            return Err(()).into();
+        }
+
+        // if any of the channels became ready then wake the endpoint up
+        if is_any_ready {
+            Poll::Ready(Ok(()))
+        } else {
+            Poll::Pending
+        }
+    }
+
+    #[inline]
+    fn queue<F: FnOnce(&mut Self::Queue)>(&mut self, f: F) {
+        let this: &'static mut Self = unsafe {
+            // Safety: As noted in the [transmute examples](https://doc.rust-lang.org/std/mem/fn.transmute.html#examples)
+            // it can be used to temporarily extend the lifetime of a reference. In this case, we
+            // don't want to use GATs until the MSRV is >=1.65.0, which means `Self::Queue` is not
+            // allowed to take generic lifetimes.
+            //
+            // We are left with using a `'static` lifetime here and encapsulating it in a private
+            // field. The `Self::Queue` struct is then borrowed for the lifetime of the `F`
+            // function. This will prevent the value from escaping beyond the lifetime of `&mut
+            // self`.
+            //
+            // See https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9a32abe85c666f36fb2ec86496cc41b4
+            //
+            // Once https://github.com/aws/s2n-quic/issues/1742 is resolved this code can go away
+            core::mem::transmute(self)
+        };
+
+        let mut capacity = 0;
+        let mut first_occupied = None;
+        for (idx, channel) in this.channels.iter_mut().enumerate() {
+            // try to make one more effort to acquire capacity for sending
```

Comment:
Yeah this is to satisfy the `capacity` function in the `io::Tx` trait as well as doing an atomic synchronization in case any more items were freed up. We can get the cached length but since we don't always poll the TX channels, it could mean that the length is out of date and lower than it could be.

Reply:
Makes sense, and sounds good.